### PR TITLE
feat: prioritize short paths during findBestHop

### PR DESCRIPTION
**Pull request by sentientwaffle**
* `findBestHopFor{Source,Destination}Amount()` should prioritize routes based on shortest `maxPathLength()`.
* Fix `maxPathLength()`: if `nextLedger` is not equal to `destinationLedger`, increment the path length.
* Possible fix for https://github.com/interledgerjs/ilp-connector/issues/382. When two paths score the same, return the first in the Map.

### CODE REVIEW

File: src/lib/routing-table.js [hunk 1]
```diff
@@ -103,30 +105,52 @@ class RoutingTable {
...
+/**
+ * If both hops score equally, return `currentHop`.
+ * It doesn't actually matter which is returned in that case, so long as it is consistent.
+ */
+function getBetterHop (currentHop, otherHop) {
+  if (!currentHop) return otherHop
+  if (currentHop.pathLength < otherHop.pathLength) return currentHop
+  if (otherHop.pathLength < currentHop.pathLength) return otherHop
+  if (otherHop.value !== undefined) {
+    return otherHop.value.gt(currentHop.value) ? otherHop : currentHop
```
Comment (sharafian) [inline src/lib/routing-table.js]: this will throw an error if currentHop.value is undefined

File: src/lib/routing-table.js [hunk 2]
```diff
@@ -103,30 +105,52 @@ class RoutingTable {
...
+  }
+  if (otherHop.cost !== undefined) {
+    return otherHop.cost.lt(currentHop.cost) ? otherHop : currentHop
```
Comment (sharafian) [inline src/lib/routing-table.js]: will throw an error if currentHop.cost is undefined

File: test/routing-table.test.js [hunk 3]
```diff
@@ -76,6 +77,16 @@ describe('RoutingTable', function () {
       const table = new RoutingTable()
       assert.strictEqual(table.findBestHopForSourceAmount(ledgerB, 10), undefined)
     })
+
+    it('prefers short routes', function () {
```
Comment (sharafian) [inline test/routing-table.test.js]: can you have a test case for each of the comparisons in getBetterHop?

File: src/lib/route.js [hunk 4]
```diff
@@ -66,7 +66,7 @@ class Route {
     // are not included in this.paths, so for instance a path:
     // * localLedger -> trustLine -> destinationLedger has this.paths === [ [] ], and path length 2
     // * localLedger -> trustLine -> anotherTrustLine -> destinationLedger has this.paths === [ ['anotherTrustLine'] ] and path length 3
-    return max + 2
+    return this.nextLedger === this.destinationLedger ? max + 2 : max + 3
```
Comment (michielbdejong) [inline src/lib/route.js]: This means we count a route `trustLine -> localLedger` as length 2, so counting the number of ledgers involved. I originally thought of it as the number of connections, so `this.nextLedger === this.destinationLedger ? max + 1 : max + 2` but I guess your way of counting makes (more) sense.

File: src/lib/routing-table.js [hunk 5]
```diff
@@ -70,29 +70,31 @@ class RoutingTable {
...
       // and get a remote quote. In the future we may refactor this
       // so that multiple next hop options can be returned, and all
       // can be asked for a quote, but for now, it's just the last.
-      if (value === undefined || !bestValue || value.gt(bestValue)) {
-        bestHop = nextHop
-        bestValue = value && value.toString()
-        bestRoute = route
-      }
+      bestHop = getBetterHop(bestHop, {
+        nextHop,
```
Comment (michielbdejong) [inline src/lib/routing-table.js]: getBetterHop takes two arguments, `function getBetterHop (currentHop, otherHop)`.
Here, you pass `bestHop` as the `currentHop`, and `{ nextHop, ... }` as the `otherHop`.
Does that mean (if both arguments of `getBetterHop` are of the same type) that `bestHop` also has a field `nextHop`? `nextHop` comes from `routes.forEach((route, nextHop) => {`, so it's a route.
Now I no longer remember what data type a 'hop' is. Is it a ledger name string? Or a Route object?

Comment (michielbdejong) [inline src/lib/routing-table.js]: Do all of `currentHop`, `otherHop`, `nextHop` and `bestHop` have the same data type? If not, then that's confusing.

File: src/lib/routing-table.js [hunk 6]
```diff
@@ -70,29 +70,31 @@ class RoutingTable {
...
+        route,
+        value: route.amountAt(sourceAmount),
+        pathLength: route.maxPathLength()
+      })
     })
 
     if (bestHop) {
-      debug('findBestHopForSourceAmount to ' + destination + ' for ' + sourceAmount + ' found route through ' + bestHop)
+      debug('findBestHopForSourceAmount to ' + destination + ' for ' + sourceAmount + ' found route through ' + bestHop.nextHop)
```
Comment (michielbdejong) [inline src/lib/routing-table.js]: See above, `bestHop.nextHop` - I don't know what that means, the next Hop of the best Hop. Which one is the Route object here and which one is the peer ledger name String?

**Pull request merged**